Update README.md with information of the model configuration file (#114)
ACCESS-NRI/model-config-tests: README.md

|/
A   (Common Ancestor)
```

### CI Configuration File

Each model configuration repository requires a `config/ci.json`

> file for defining CI test configuration. This file specifies what scheduled tests to run, and what tests to run for a given git branch of tag and test type.

Reviewer: ```suggestion
file for defining CI test configuration. This file specifies what scheduled tests to run, and what tests to run for a given git branch (or tag) and test type.
```

Reviewer: I think that is what was intended, so just tried to make it clearer.

Author: Thanks!

#### Test types

The different types of test are defined as:

> - `scheduled`: Scheduled monthly reproducibility tests run on NCI. The keys under these tests are released configuration tags to run the scheduled tests.

Author: Yeah true, it can run for git branches too. I've also added a note that it's typically scheduled monthly but can be reconfigured.

> - `reproducibility`: Reproducibility tests run on NCI as part of pull requests. The keys under these tests represent the target branches into which pull requests are being merged. These are automatically run for pull requests from development (`dev-`) branches to released (`release-`) branches. These tests can also be triggered manually in a PR using the `!test repro` command.

Author: Ok I've committed to just using just one `Pull Requests (PRs)` and the rest `PRs`

> - `reproducibility`: Reproducibility tests run as part of pull requests and are run on NCI. The keys under these tests represent the target branches into which pull requests are being merged. These are automatically run for pull requests from development (`dev-`) branches to released (`release-`) branches. These tests can also be triggered manually in a PR using the `!test repro` command.

Reviewer: ```suggestion
- `reproducibility`: Reproducibility tests run on NCI as part of pull requests. The keys under these tests represent the target branches into which pull requests are being merged. These are automatically run for pull requests from development (`dev-`) branches to released (`release-`) branches. These tests can also be triggered manually in a PR using the `!test repro` command.
```

> - `qa` - Quick quality assurance tests are run locally on a GitHub Runner as part of pull requests . The keys under these tests represent the target branches into which pull requests are being merged. These are run for any PR being merged into development or released branches.

Author: Yeah it worth making that clear - I've added that in

> - `qa` - Quick quality assurance tests are run as part of pull requests and run locally on a GitHub Runner. The keys under these tests represent the target branches into which pull requests are being merged. These are run for any PR being merged into development or released branches.

Reviewer: ```suggestion
- `qa` - Quick quality assurance tests are run locally on a GitHub Runner as part of pull requests . The keys under these tests represent the target branches into which pull requests are being merged. These are run for any PR being merged into development or released branches.
```

| Name | Type | Description |  Example |
| ---- | ---- | ----------- | -------- |
| markers | `string` | Markers used for the pytest checks, in the Python format | `checksum` |
| model-config-tests-version | `string` | The version of the model-config-tests | `0.0.1` |
| python-version | `string` | The python version used to create test virtual environment on GitHub hosted tests | `3.11.0` |
| payu-version | `string` | The Payu version used to run the model | `1.1.5` |

Pytest markers select what pytests to run in `model-config-tests`. Current markers include:

- `checksum`: Reproducibility test that runs a model and compares checksums with a stored previous result.
- `checksum_slow`: Slower reproducibility tests. These include repeating running a configuration from initial conditions to check it reproduces the same output, and checking restart reproducibility by comparing two shorter model runs with a longer model run.
- `dev_config`: General configuration QA tests.
- `config`: Configuration QA tests for released branches.

There are also model-specific markers for configuration QA tests, e.g., `access_om2`, `access_esm1p5`, `access_om3` and `access_esm1p6`.

> To select a combination of tests, use the `or` keyword, e.g. `"markers": "dev_config or config"`.

Reviewer 2: ```suggestion
To select a combination of tests, use the `or` keyword, e.g. `"markers": "access_om2 or config"`.
```

> The Payu version is the module version loaded on NCI to build the base of the test virtual environment. To use the development Payu module which has all the latest changes in Payu repository, set `"payu-version": "dev"`.

Reviewer: ```suggestion
The Payu version is the module version loaded on NCI to build the base of the test virtual environment and is only relevant for tests run on NCI. To use the development Payu module which has all the latest changes in Payu repository, set `"payu-version": "dev"`.
```

> For development model configuration branches, it may be useful to use `main` branch of the `model-config-tests` repository. A git branch, tag, or commit of the `model-config-tests` repository can be used for `model-config-tests-version`, and the CI will install from Github if the package can't be installed from PyPI. Similarly, with the development Payu version, it is recommended to use released versions so the released configurations tests are stable.

Reviewer: ```suggestion
A git branch, tag, or commit of the `model-config-tests` repository can be used for `model-config-tests-version`, and the CI will install from Github if the package can't be installed from PyPI. For development model configuration branches it may be useful to use the most recent code in the `model-config-tests` repository. Using the `main` branch will achieve this. This is fine for testing, however it is recommended to use released versions of `model-config-tests` and `payu` so the released configurations tests are stable. 
```

> A git branch, tag, or commit of the `model-config-tests` repository can be used for `model-config-tests-version`, and the CI will install from Github if the package can't be installed from PyPI. For development model configuration branches it may be useful to use the most recent code in the `model-config-tests` repository. Using the `main` branch will achieve this. This is fine for testing, however it is recommended to use released versions of `model-config-tests` and `payu` so the released configurations tests are stable. 

Reviewer 2: ```suggestion
A git branch, tag, or commit of the `model-config-tests` repository can be used for `model-config-tests-version`, and the CI will install from Github if the package can't be installed from PyPI. For development model configuration branches it may be useful to use the most recent code in the `model-config-tests` repository. Using the `main` branch will achieve this. This is good for continuous integration, however it is recommended to use released versions of `model-config-tests` and `payu` for the released model configurations so that tests are stable. 
```

There is also support for using regex patterns for the git branches for `qa` and `reproducibility` tests, e.g.
`dev-.*` will match all development branches. If there are multiple regex patterns, priority will be given to the order so more specific patterns should
be higher in the list, e.g. `dev-1deg-.*` before `dev-.*`. Note that regex patterns are currently not supported for `scheduled` tests so every tag that requires a scheduled test needs to be defined.

For example, given the following `config/ci.json` file:

```json
{
    "$schema": "<https://github.com/ACCESS-NRI/schema/tree/main/au.org.access-nri/model/configuration/ci/2-0-0.json>",
    "scheduled": {
        "release-1deg_jra55_ryf-2.0": {},
        "default": {
            "markers": "checksum"
        },
    },
    "reproducibility": {
        "default": {
            "markers": "checksum"
        },
        "dev-1deg_jra55do_ryf": {
            "markers": "checksum or checksum_slow"
        },
        "release-1deg_jra55do_ryf": {
            "markers": "checksum or checksum_slow"
        },
        "dev-example-branch": {
            "payu-version": "dev",
            "model-config-tests-version": "main"
        },

>     },

Reviewer 3: Should the reproducibility example also have `default` at the bottom like the others? Is that convention, or required?

Author: Yeah it doesn't need to be ordered at the bottom but I'll add it in [a028d20](https://github.com/ACCESS-NRI/model-config-tests/pull/114/commits/a028d2053388d2ae2303d25baa502c8b41c51f14) so it looks neater.

The different types of test are defined as:

> - `scheduled`: Scheduled reproducibility tests run on NCI. These are typically scheduled to run monthly but this can re-configured by modifying the cron in the `.github/workflows/schedule.yml` workflow. The keys under these tests are released configuration tags or branches to run the scheduled tests.

Reviewer 2: ```suggestion
- `scheduled`: Scheduled reproducibility tests run on NCI. These are typically scheduled to run monthly but this can re-configured by modifying the cron in the `.github/workflows/schedule.yml` workflow. The keys under these tests are released tags or branches to run the scheduled tests.
```
